### DISCUSSION TITLE
incremental displacements at the boundaries using nodal coordinates

### DISCUSSION THREAD

**Discussion by gsoim** [idaholab/moose#23152] [Q&A Modules: Solid mechanics] [answered]
Hi everyone,

I am trying to solve a plane stress elasto-plastic problem (with J2 plasticity) using incremental displacements at the boundaries. To do so, I have four parameters F11, F12, F21, F22 that need to be multiplied with the nodal coordinates (x,y) of the undeformed mesh at the specific boundary.

For example, the four parameters are defined in the input file as:

```
F11 = 0.01
F12 = 0.01
F21 = 0.01
F22 = 0.01
```

while the Function for the right boundary is defined as:

```
[Functions]
  # [./right_disp_x]
  #   type = PiecewiseLinear
  #   x = '0    1.0    2.0    3.0     4.0    5.0    6.0    7.0    8.0    9.0   10.0'
  #   y = '0  0.003  0.006  0.009   0.012  0.015  0.018  0.021  0.024  0.027  0.030'
  # [../]
  [./right_disp_x]
    type = PiecewiseLinear
    total_disp_x = ${fparse ${raw F11}*x + ${raw F12}*y - x}  # (F11*x + F12*y) - x
    dx = ${fparse ${raw total_disp_x} / 10}
    x = '0    1.0    2.0    3.0     4.0    5.0    6.0    7.0    8.0    9.0   10.0'
    y = '0     dx   2*dx   3*dx    4*dx   5*dx   6*dx   7*dx   8*dx   9*dx  10*dx'
  [../]
[]
```

In the above snippet I have commented out the simple Function that imposes the displacement at the boundary using 
specific numerical values.

And finally, the BCs are defined as following:

```
[BCs]
  [./bottom_y]
    type = DirichletBC
    boundary = 'bottom'
    variable = disp_y
    value = 0.0
  [../]
  [./left_x]
    type = DirichletBC
    boundary = 'left'
    variable = disp_x
    value = 0.0
  [../]
  [./right_x]
    type = FunctionDirichletBC
    boundary = 'right'
    variable = disp_x
    function = right_disp_x
  [../]
[]
```

Unfortunately, I get the following error:

*** ERROR ***
/Users/georgiossoimoiris/projects/plane_stress/problems/plane_stress_plastic_DefGradBC.i:202: field node 'Functions/right_disp_x/x' does not hold a float-typed value (val=''0    1.0    2.0    3.0     4.0    5.0    6.0    7.0    8.0    9.0   10.0'')
/Users/georgiossoimoiris/projects/plane_stress/problems/plane_stress_plastic_DefGradBC.i:203: field node 'Functions/right_disp_x/total_disp_x' does not hold a float-typed value (val='${fparse ${raw F11}*x + ${raw F12}*y - x}')

which means that 'x' and 'y' are not the nodal coordinates that I need to use!

So, my question is: how can I get the nodal coordinates at a specific boundary and multiply them as a vector with my parameters in order to add them to my BCs as a Function incrementally??

Thank you very much,
George

**Comment by GiudGiud** [comment, marked as answer]
Hello

Indeed the format of these parameters are not quite right. 

You can write a parsedFunction:
```
  [./bc_func]
    type = ParsedFunction
    expression = 'F11 * x + F12 * y'
    symbol_names = 'F11 F12'
    symbol_values = '${F11} ${F12}'
  [../]
```

then use the FunctionDirichletBC like you did

Guillaume

**Reply by gsoim** [reply]
Hello,

Thanks for the response.
Using the above I got the following error:

```
*** ERROR ***
/Users/.../plane_stress_plastic_DefGradBC.i:202: missing required parameter 'Functions/right_disp_x/value'
	Doc String: "The user defined function."
```

Why I get this value error??

Moreover, since I want to impose these displacements incrementally, would it possible to multiply it with time "t" like that:

```
[./right_disp_x]
    type = ParsedFunction
    expression = '((F11 * x + F12 * y) - y) * t'  # instead of:  'F11 * x + F12 * y'
    symbol_names = 'F11 F12'
    symbol_values = '${F11} ${F12}'
[../]
```
Since I am using a transient  solver like that:

```
[Executioner]
  type = Transient
  end_time = 10
  dt = 1

  solve_type = 'newton'
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'

  # controls for linear iterations
  l_max_its = 100
  l_tol = 1e-03
  
  # controls for nonlinear iterations
  nl_max_its = 100
  nl_rel_tol = 1e-03
  nl_abs_tol = 1e-03
[]
```
Thanks,
George

**Reply by GiudGiud** [reply]
Your version is not up to date. This parameter got renamed from value to expression

**Reply by gsoim** [reply]
I updated MOOSE and now it seems to be working quite fine.
Thanks!
